### DISCUSSION TITLE
Custom Trivy-Secrets rule || Bug

### DISCUSSION THREAD

**Discussion by suspect22** [aquasecurity/trivy#4617] [Q&A] [answered]
### Question

Currently I am working on a Custom Ruleset for Appsettings in dotnet Core.
We are using Trivy for Image and Vulnerability Scanning. We also wanted to utilize it for Secret Scanning therefore we implemented some Custom Rules.

The rule i have written is not matching.

## File Example is
`
{
  "Serilog": {
    "MinimumLevel": {
      "Default": "Information",
      "Override": {
        "Microsoft": "Warning",
        "Microsoft.Hosting.Lifetime": "Information"
      }
    }
  },
  "ApiKey": {
   "Value": "2153b939-5b5a-46e9-a2cc-15c34834a10e"
  },
  "FileSystemBlobStore":  {
    "StorageDirectory": "/tmp/blob"
  },
  "AllowedHosts": "*"
}
`

## Rule example is
`
rules:
  - id: Test for Keys in this Project
    category: Passwords
    title: API Key Issue
    severity: HIGH
    path: .\\.json
    keywords:
      - secret
    regex: (?i)(?:["'])(?P<key>Value)(?:["':]*)\s*(?:["'])(?P<secret>[^"]*)(?:["'])
    secret-group-name: secret
    allow-rules:
      - id: Development Appsettings
        description: Skip Development Appsettings File
        path: .*\appsettings.Development.json
      - id: replaced by
        description: Skip if replaced by is in the Secret
        regex: replaced_by_environment
`

## Tested against https://regex101.com/

![grafik](https://github.com/aquasecurity/trivy/assets/7037695/5c810fc7-1369-46bc-bc8d-7aa225a8c13a)

**Trivy is not finding the Secret**, is this a user related/Regex Error?
Am I doing something wrong? What am I missing ;)

Thanks for your help
Kind regards

suspect22

### Target

Filesystem

### Scanner

Secret

### Output Format

None

### Mode

Standalone

### Operating System

5.15.90.1-microsoft-standard-WSL2 #1 SMP Fri Jan 27 02:56:13 UTC 2023 x86_64 x86_64 x86_64 GNU/Linu

### Version

```bash
Version: 0.42.1
Vulnerability DB:
  Version: 2
  UpdatedAt: 2023-06-02 12:07:36.18134768 +0000 UTC
  NextUpdate: 2023-06-02 18:07:36.18134738 +0000 UTC
  DownloadedAt: 2023-06-02 12:15:39.45102896 +0000 UTC
Java DB:
  Version: 1
  UpdatedAt: 2023-05-26 00:54:30.554694094 +0000 UTC
  NextUpdate: 2023-05-29 00:54:30.554693594 +0000 UTC
  DownloadedAt: 2023-05-26 14:59:41.078265016 +0000 UTC
Policy Bundle:
  Digest: sha256:030345bff4845bae1c89450da85681bc25c0c40d2144a0d9db5254674d155298
  DownloadedAt: 2023-06-02 10:34:55.99769366 +0000 UTC
```

**Comment by suspect22** [comment]
Anyone has a clue about that?

**Reply by DmitriyLewen** [reply, marked as answer]
Hello @suspect22 
Sorry for late reply.

I checked you issue and found 2 cases:
1) `path: .\.json`. You need to use regex for this field - https://aquasecurity.github.io/trivy/v0.42/docs/scanner/secret/#custom-rules
2) you `regex` field is not equal with screenshot.

I updated your secret-config file(fixed `path` and took `regex` from screenshot):
```
➜  cat trivy-secret.yaml 
rules:
  - id: Test for Keys in this Project
    category: Passwords
    title: API Key Issue
    severity: HIGH
    path: .*\.json
    secret-group-name: secret
    regex: (?i)(?:["'])(?P<key>Value)(?:["':]*)\s*(?:["'])(?P<secret>[^"]*)(?:["'])

➜  cat 123.json
{
  "Serilog": {
    "MinimumLevel": {
      "Default": "Information",
      "Override": {
        "Microsoft": "Warning",
        "Microsoft.Hosting.Lifetime": "Information"
      }
    }
  },
  "ApiKey": {
    "Value": "2153b939-5b5a-46e9-a2cc-15c34834a10e"
  },
  "FileSystemBlobStore": {
    "StorageDirectory": "/tmp/blob"
  },
  "AllowedHosts": "*"
}

➜  trivy fs --scanners secret --secret-config ./trivy-secret.yaml .       
2023-06-30T11:59:59.986+0600	INFO	Secret scanning is enabled
2023-06-30T11:59:59.986+0600	INFO	If your scanning is slow, please try '--scanners vuln' to disable secret scanning
2023-06-30T11:59:59.986+0600	INFO	Please see also https://aquasecurity.github.io/trivy/v0.42/docs/secret/scanning/#recommendation for faster secret detection
2023-06-30T11:59:59.986+0600	INFO	Loading ./trivy-secret.yaml for secret scanning...

123.json (secrets)

Total: 1 (UNKNOWN: 0, LOW: 0, MEDIUM: 0, HIGH: 1, CRITICAL: 0)

HIGH: Passwords (Test for Keys in this Project)
════════════════════════════════════════════════════════════════════════════════
API Key Issue
────────────────────────────────────────────────────────────────────────────────
 123.json:12
────────────────────────────────────────────────────────────────────────────────
  10     },
  11     "ApiKey": {
  12 [     "Value": "************************************"
  13     },
────────────────────────────────────────────────────────────────────────────────
```

PS: don't use `test` in dir/file names, we skip these files - https://github.com/aquasecurity/trivy/blob/22463ababd289efaf7121db9f83047223ef43bab/pkg/fanal/secret/builtin-allow-rules.go#L4-L8
